File: tools/py/src/leyline_tools/arena/hand.py
from __future__ import annotations

import json
import math
import re
from pathlib import Path

from PIL import Image

from .common import die, try_remove
from .debug_api import fetch_api
from .macos import capture_window, ocr
from .paths import REFERENCE_WIDTH
from .scry_bridge import _scry_state
from .shell import run


HAND_MIN_CY = 490

_HAND_Y_RATIO = 0.78
_HAND_CY_CENTER = 530
_HAND_ARC_DROP = 35
_HAND_X_MIN = 120
_HAND_X_MAX = 800
_HAND_X_CENTER = 500
_HAND_CX_BASE_SHIFT = 20
_HAND_CX_NUDGE = 20
_OCR_TARGET_WIDTH = 3840


def cmd_ocr_hand() -> None:
    known_names: list[str] = []
    state = _scry_state()
    if state:
        known_names = [
            c.get("name", "") for c in state.get("hand", []) if c.get("name")
        ]

    if not known_names:
        raw = fetch_api("/api/id-map?active=true")
        if raw:
            entries = json.loads(raw)
            if isinstance(entries, dict):
                entries = entries.get("data", [])
            known_names = [
                e.get("cardName", "")
                for e in entries
                if e.get("forgeZone") == "Hand"
                and e.get("ownerSeatId") == 1
                and e.get("cardName")
            ]

    if not known_names:
        die("No hand cards found (scry and debug API both empty)")

    result = _ocr_hand_strip()
    if result is None:
        die("Hand OCR failed (capture or OCR returned no results)")
    items, ocr_to_960 = result

    print(f"Known hand ({len(known_names)}): {known_names}")
    print(f"OCR detections ({len(items)}):")
    for item in items:
        cx_960 = int(item["cx"] * ocr_to_960)
        print(
            f"  {item['text']:30s} cx={cx_960:4d}  conf={item.get('confidence', 0):.2f}"
        )

    matched: dict[str, tuple[float, int]] = {}
    for item in items:
        cx_960 = int(item["cx"] * ocr_to_960)
        if cx_960 < _HAND_X_MIN or cx_960 > _HAND_X_MAX:
            continue
        for name in known_names:
            score = _fuzzy_card_match(item["text"], name)
            if score >= 0.4:
                prev = matched.get(name)
                if prev is None or score > prev[0]:
                    matched[name] = (score, cx_960)

    print(f"\nMatched ({len(matched)}/{len(known_names)}):")
    for name in known_names:
        if name in matched:
            score, cx = matched[name]
            print(f"  {name:30s} cx={cx:4d}  score={score:.2f}")
        else:
            print(f"  {name:30s} NOT FOUND")


def _hand_adjust(ocr_cx: int) -> tuple[int, int]:
    offset_ratio = (_HAND_X_CENTER - ocr_cx) / (_HAND_X_CENTER - _HAND_X_MIN)
    cx = ocr_cx + _HAND_CX_BASE_SHIFT + int(_HAND_CX_NUDGE * offset_ratio)
    dx = (cx - _HAND_X_CENTER) / (_HAND_X_MAX - _HAND_X_MIN)
    arc_offset = _HAND_ARC_DROP * (1 - math.cos(dx * math.pi))
    cy = _HAND_CY_CENTER + int(arc_offset)
    return (cx, cy)


def _ocr_upscale_factor(capture_px_width: int) -> int:
    if capture_px_width <= 0:
        return 2
    return max(2, round(_OCR_TARGET_WIDTH / capture_px_width))


def _ocr_hand_strip() -> tuple[list[dict], float] | None:
    tmp = "/tmp/arena/_hand_ocr"
    Path(tmp).mkdir(parents=True, exist_ok=True)

    img = f"{tmp}/capture.png"
    bounds = capture_window(img, hires=True)
    if bounds is None:
        return None

    _, sips_out, _ = run("sips", "-g", "pixelWidth", "-g", "pixelHeight", img)
    w_m = re.search(r"pixelWidth:\s*(\d+)", sips_out)
    h_m = re.search(r"pixelHeight:\s*(\d+)", sips_out)
    if not w_m or not h_m:
        try_remove(img)
        return None

    cap_w = int(w_m.group(1))
    cap_h = int(h_m.group(1))

    scale = _ocr_upscale_factor(cap_w)
    crop_top = int(cap_h * _HAND_Y_RATIO)
    pil_img = Image.open(img)
    strip_pil = pil_img.crop((0, crop_top, cap_w, cap_h))
    strip_up = f"{tmp}/strip_up.png"
    strip_pil = strip_pil.resize(
        (cap_w * scale, strip_pil.height * scale), Image.LANCZOS
    )
    strip_pil.save(strip_up)
    pil_img.close()
    strip_pil.close()
    try_remove(img)

    code, stdout, _ = ocr(strip_up, "--min-confidence", "0.15")
    try_remove(strip_up)
    if code != 0 or not stdout.strip():
        return None

    items = json.loads(stdout)
    ocr_to_960 = REFERENCE_WIDTH / (cap_w * scale)
    return (items, ocr_to_960)


def _fuzzy_card_match(ocr_text: str, card_name: str) -> float:
    ot = ocr_text.lower().strip()
    for ch in "''\u2019()[].,":
        ot = ot.replace(ch, "")
    ot = ot.strip()
    cn = card_name.lower().strip().replace("'", "").replace("\u2019", "")

    if not ot or not cn:
        return 0.0
    if ot == cn:
        return 1.0
    if cn in ot:
        return 0.9
    if ot in cn and len(ot) >= 3:
        return 0.7 * len(ot) / len(cn)

    best = 0.0
    cn_words = cn.split()
    ot_words = ot.split()
    if len(cn_words) > 1 and ot_words:
        hits = 0
        for w in cn_words:
            if len(w) <= 2:
                continue
            for ow in ot_words:
                if w in ow or ow in w or _levenshtein(w, ow) <= 1:
                    hits += 1
                    break
        if hits > 0:
            best = max(best, 0.6 * hits / len(cn_words))

    if len(cn) <= 8 and len(ot) <= 12:
        dist = _levenshtein(ot, cn)
        if dist <= 1:
            best = max(best, 0.8)
        elif dist <= 2 and len(cn) >= 5:
            best = max(best, 0.5)

    if abs(len(ot) - len(cn)) <= 3 and len(cn) >= 6:
        dist = _levenshtein(ot, cn)
        if dist <= 2:
            best = max(best, 0.7)
        elif dist <= 3 and len(cn) >= 10:
            best = max(best, 0.5)

    return best


def _levenshtein(a: str, b: str) -> int:
    if len(a) < len(b):
        return _levenshtein(b, a)
    if not b:
        return len(a)
    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a):
        curr = [i + 1]
        for j, cb in enumerate(b):
            curr.append(
                min(prev[j + 1] + 1, curr[j] + 1, prev[j] + (0 if ca == cb else 1))
            )
        prev = curr
    return prev[-1]


def _find_hand_card_ocr(
    card_name: str, known_names: list[str]
) -> tuple[int, int] | None:
    result = _ocr_hand_strip()
    if result is None:
        return None

    items, ocr_to_960 = result
    if not items:
        return None

    card_matches: dict[str, list[tuple[float, int]]] = {}
    for item in items:
        cx_960 = int(item["cx"] * ocr_to_960)
        if cx_960 < _HAND_X_MIN or cx_960 > _HAND_X_MAX:
            continue
        for name in known_names:
            score = _fuzzy_card_match(item["text"], name)
            if score >= 0.4:
                card_matches.setdefault(name, []).append((score, cx_960))

    resolved: dict[str, int] = {}
    used_cx: set[int] = set()
    for name in sorted(
        card_matches, key=lambda n: max(s for s, _ in card_matches[n]), reverse=True
    ):
        candidates = sorted(card_matches[name], key=lambda t: -t[0])
        for score, cx in candidates:
            if not any(abs(cx - ux) < 30 for ux in used_cx):
                resolved[name] = cx
                used_cx.add(cx)
                break
        if name not in resolved and candidates:
            resolved[name] = candidates[0][1]

    target_lower = card_name.lower()
    for name, ocr_cx in resolved.items():
        if name.lower() == target_lower or target_lower in name.lower():
            return _hand_adjust(ocr_cx)

    if len(resolved) == len(known_names) - 1:
        count = len(known_names)
        spacing = min(80, 400 // max(count, 1))
        all_positions = [
            480 - (count - 1) * spacing // 2 + i * spacing for i in range(count)
        ]
        for pos in all_positions:
            if not any(abs(pos - ocr_cx) < 30 for ocr_cx in resolved.values()):
                return _hand_adjust(pos)

    return None


def _find_hand_card(name: str) -> tuple[tuple[int, int], int] | None:
    hand_cards: list[dict] = []
    state = _scry_state()
    if state:
        for card in state.get("hand", []):
            hand_cards.append(
                {
                    "instanceId": card.get("id"),
                    "cardName": card.get("name", ""),
                }
            )

    if not hand_cards:
        return None

    target = None
    name_lower = name.lower()
    for card in hand_cards:
        card_name = (card.get("cardName") or "").lower()
        if card_name == name_lower or name_lower in card_name:
            target = card
            break
    if target is None:
        return None

    instance_id = target.get("instanceId", 0)
    known_names = [c.get("cardName") or "" for c in hand_cards if c.get("cardName")]
    ocr_pos = _find_hand_card_ocr(name, known_names)
    if ocr_pos is not None:
        return (ocr_pos, instance_id)

    # Fallback: detection model (hand-card bounding boxes)
    from .board import _board_detect

    idx = next(
        (i for i, c in enumerate(hand_cards) if c.get("instanceId") == instance_id),
        0,
    )
    dets = _board_detect()
    hand_dets = sorted(
        [d for d in dets if d["label"] == "hand-card" and d["cy"] > HAND_MIN_CY],
        key=lambda d: d["cx"],
    )

    if hand_dets:
        if len(hand_dets) == len(hand_cards) and idx < len(hand_dets):
            det = hand_dets[idx]
        elif idx < len(hand_dets):
            det = hand_dets[idx]
        else:
            det = hand_dets[-1]
        return ((det["cx"], det["cy"]), instance_id)

    # Last resort: estimate position from card count
    count = len(hand_cards)
    spacing = min(80, 400 // max(count, 1))
    center_x = 480 - (count - 1) * spacing // 2 + idx * spacing
    return (_hand_adjust(center_x), instance_id)

File: tools/arena-ts/src/hand.ts
// src/hand.ts
// Hand card detection via OCR.
// Captures the hand strip of the MTGA window, upscales for OCR accuracy,
// matches detected text against known card names from scry game state.

import { captureMtga } from "./input";
import { compileSwift } from "./compile";
import { REFERENCE_WIDTH } from "./window";

// Hand layout constants (960px reference space)
const HAND_CROP_BOTTOM = 0.20;   // capture bottom 20% of window
const HAND_CROP_SIDES = 280;     // trim 280px (retina) from each side
const HAND_X_MIN = 120;          // left bound for valid hand card detections
const HAND_X_MAX = 800;          // right bound
const HAND_X_CENTER = 500;       // visual center of hand fan
const HAND_CY_CENTER = 530;      // y center of hand cards
const HAND_ARC_DROP = 35;        // arc curvature (cards at edges sit lower)
const HAND_CX_BASE_SHIFT = 20;   // OCR text sits left of card center
const HAND_CX_NUDGE = 20;        // extra shift for edge cards
const OCR_TARGET_WIDTH = 3840;   // upscale target for better OCR

export interface HandCardPosition {
  name: string;
  cx: number;     // 960px reference x
  cy: number;     // 960px reference y
  score: number;  // match confidence 0-1
}

/** Adjust OCR-detected x to actual card center, accounting for arc. */
function handAdjust(ocrCx: number): [number, number] {
  const offsetRatio = (HAND_X_CENTER - ocrCx) / (HAND_X_CENTER - HAND_X_MIN);
  const cx = ocrCx + HAND_CX_BASE_SHIFT + Math.round(HAND_CX_NUDGE * offsetRatio);
  const dx = (cx - HAND_X_CENTER) / (HAND_X_MAX - HAND_X_MIN);
  const arcOffset = HAND_ARC_DROP * (1 - Math.cos(dx * Math.PI));
  const cy = HAND_CY_CENTER + Math.round(arcOffset);
  return [cx, cy];
}

/**
 * Capture MTGA hand strip, upscale, OCR.
 * Pipeline: capture by window ID → crop (bottom 20%, trim 280px sides) → upscale → OCR.
 * Works even if MTGA is behind other windows.
 */
async function ocrHandStrip(): Promise<{ items: any[]; } | null> {
  const fullImg = "/tmp/arena-hand-full.png";
  const stripImg = "/tmp/arena-hand-strip.png";
  const stripUp = "/tmp/arena-hand-strip-up.png";

  // Capture by window ID
  if (!(await captureMtga(fullImg))) return null;

  // Get pixel dimensions
  const sipsInfo = Bun.spawnSync({ cmd: ["sips", "-g", "pixelWidth", "-g", "pixelHeight", fullImg], stdout: "pipe" });
  const sipsOut = sipsInfo.stdout.toString();
  const wMatch = sipsOut.match(/pixelWidth:\s*(\d+)/);
  const hMatch = sipsOut.match(/pixelHeight:\s*(\d+)/);
  if (!wMatch || !hMatch) return null;
  const fullW = parseInt(wMatch[1]);
  const fullH = parseInt(hMatch[1]);

  // Crop: bottom 20%, trim sides via Swift CGImage.cropping
  const cropH = Math.round(fullH * HAND_CROP_BOTTOM);
  const cropY = fullH - cropH;
  const cropX = HAND_CROP_SIDES;
  const cropW = fullW - HAND_CROP_SIDES * 2;

  const cropProc = Bun.spawnSync({
    cmd: ["swift", "-e", `
import AppKit
let img = NSImage(contentsOfFile: "${fullImg}")!
let cg = img.cgImage(forProposedRect: nil, context: nil, hints: nil)!
let cropped = cg.cropping(to: CGRect(x: ${cropX}, y: ${cropY}, width: ${cropW}, height: ${cropH}))!
let rep = NSBitmapImageRep(cgImage: cropped)
try! rep.representation(using: .png, properties: [:])!.write(to: URL(fileURLWithPath: "${stripImg}"))
`],
    stderr: "pipe",
  });
  if (cropProc.exitCode !== 0) return null;

  // Upscale strip for OCR (target ~3840px wide)
  const scale = Math.max(2, Math.round(3840 / cropW));
  Bun.spawnSync({ cmd: ["sips", "--resampleWidth", String(cropW * scale), stripImg, "--out", stripUp] });

  // OCR
  const ocrBin = await compileSwift("ocr");
  const ocr = Bun.spawnSync({
    cmd: [ocrBin, stripUp, "--json", "--min-confidence", "0.10"],
    stdout: "pipe",
  });
  if (ocr.exitCode !== 0) return null;

  try {
    const items = JSON.parse(ocr.stdout.toString());
    // Convert OCR coords (upscaled-strip space) → 960px ref space
    for (const item of items) {
      item.cx = (item.cx / scale + cropX) / fullW * REFERENCE_WIDTH;
    }
    return { items }; // cx already in 960px ref
  } catch {
    return null;
  }
}

/** Levenshtein distance between two strings. */
function levenshtein(a: string, b: string): number {
  if (a.length < b.length) return levenshtein(b, a);
  if (b.length === 0) return a.length;
  let prev = Array.from({ length: b.length + 1 }, (_, i) => i);
  for (let i = 0; i < a.length; i++) {
    const curr = [i + 1];
    for (let j = 0; j < b.length; j++) {
      curr.push(Math.min(
        prev[j + 1] + 1,
        curr[j] + 1,
        prev[j] + (a[i] === b[j] ? 0 : 1),
      ));
    }
    prev = curr;
  }
  return prev[b.length];
}

/** Fuzzy match OCR text against a card name. Returns 0-1 score. */
function fuzzyCardMatch(ocrText: string, cardName: string): number {
  let ot = ocrText.toLowerCase().replace(/['''\u2019()\[\].,]/g, "").trim();
  let cn = cardName.toLowerCase().replace(/['''\u2019]/g, "").trim();

  if (!ot || !cn) return 0;
  if (ot === cn) return 1.0;
  if (cn.includes(ot) && ot.length >= 3) return 0.9;
  if (ot.includes(cn)) return 0.9;
  if (ot.length >= 3 && cn.includes(ot)) return 0.7 * ot.length / cn.length;

  let best = 0;

  // Word-level matching
  const cnWords = cn.split(/\s+/);
  const otWords = ot.split(/\s+/);
  if (cnWords.length > 1 && otWords.length > 0) {
    let hits = 0;
    for (const w of cnWords) {
      if (w.length <= 2) continue;
      for (const ow of otWords) {
        if (w.includes(ow) || ow.includes(w) || levenshtein(w, ow) <= 1) {
          hits++;
          break;
        }
      }
    }
    if (hits > 0) best = Math.max(best, 0.6 * hits / cnWords.length);
  }

  // Short name exact-ish match
  if (cn.length <= 8 && ot.length <= 12) {
    const dist = levenshtein(ot, cn);
    if (dist <= 1) best = Math.max(best, 0.8);
    else if (dist <= 2 && cn.length >= 5) best = Math.max(best, 0.5);
  }

  // Similar length fuzzy
  if (Math.abs(ot.length - cn.length) <= 3 && cn.length >= 6) {
    const dist = levenshtein(ot, cn);
    if (dist <= 2) best = Math.max(best, 0.7);
    else if (dist <= 3 && cn.length >= 10) best = Math.max(best, 0.5);
  }

  return best;
}

/**
 * Find a specific card in the hand by name using OCR.
 * Returns position in 960px reference space, or null if not found.
 */
export async function findHandCard(
  cardName: string,
  knownNames: string[],
): Promise<HandCardPosition | null> {
  const result = await ocrHandStrip();
  if (!result) return null;

  const { items } = result;

  // Match all OCR detections against all known hand card names
  const cardMatches = new Map<string, { score: number; cx: number }[]>();
  for (const item of items) {
    const cx960 = Math.round(item.cx);
    if (cx960 < HAND_X_MIN || cx960 > HAND_X_MAX) continue;

    for (const name of knownNames) {
      const score = fuzzyCardMatch(item.text, name);
      if (score >= 0.4) {
        if (!cardMatches.has(name)) cardMatches.set(name, []);
        cardMatches.get(name)!.push({ score, cx: cx960 });
      }
    }
  }

  // Resolve: assign best non-overlapping position to each card
  const resolved = new Map<string, number>();
  const usedCx = new Set<number>();

  const sortedNames = [...cardMatches.keys()].sort((a, b) => {
    const aMax = Math.max(...cardMatches.get(a)!.map(m => m.score));
    const bMax = Math.max(...cardMatches.get(b)!.map(m => m.score));
    return bMax - aMax;
  });

  for (const name of sortedNames) {
    const candidates = cardMatches.get(name)!.sort((a, b) => b.score - a.score);
    for (const { cx } of candidates) {
      if (![...usedCx].some(ux => Math.abs(cx - ux) < 30)) {
        resolved.set(name, cx);
        usedCx.add(cx);
        break;
      }
    }
    // Fallback: use best match even if overlapping
    if (!resolved.has(name) && candidates.length > 0) {
      resolved.set(name, candidates[0].cx);
    }
  }

  // Find target card
  const targetLower = cardName.toLowerCase();
  for (const [name, ocrCx] of resolved) {
    if (name.toLowerCase() === targetLower || name.toLowerCase().includes(targetLower) || targetLower.includes(name.toLowerCase())) {
      const [cx, cy] = handAdjust(ocrCx);
      const matches = cardMatches.get(name) ?? [];
      const bestScore = matches.length > 0 ? Math.max(...matches.map(m => m.score)) : 0;
      return { name, cx, cy, score: bestScore };
    }
  }

  // Gap inference: if we found N-1 of N cards, the missing one is at the gap
  if (resolved.size === knownNames.length - 1) {
    const count = knownNames.length;
    const spacing = Math.min(80, Math.floor(400 / Math.max(count, 1)));
    const positions = Array.from({ length: count }, (_, i) =>
      480 - Math.floor((count - 1) * spacing / 2) + i * spacing
    );
    for (const pos of positions) {
      if (![...resolved.values()].some(rx => Math.abs(pos - rx) < 30)) {
        const [cx, cy] = handAdjust(pos);
        return { name: cardName, cx, cy, score: 0.3 };
      }
    }
  }

  return null;
}

/**
 * Find positions for ALL hand cards in one OCR pass.
 * Handles duplicate card names by assigning distinct positions.
 */
export async function findAllHandCards(
  knownNames: string[],
): Promise<Map<string, HandCardPosition[]>> {
  const resultMap = new Map<string, HandCardPosition[]>();
  const ocrResult = await ocrHandStrip();
  if (!ocrResult) return resultMap;

  const { items, ocrTo960 } = ocrResult;

  // Match OCR detections to card names
  const cardMatches = new Map<string, { score: number; cx: number }[]>();
  for (const item of items) {
    const cx960 = Math.round(item.cx);
    if (cx960 < HAND_X_MIN || cx960 > HAND_X_MAX) continue;
    for (const name of new Set(knownNames)) {
      const score = fuzzyCardMatch(item.text, name);
      if (score >= 0.4) {
        if (!cardMatches.has(name)) cardMatches.set(name, []);
        cardMatches.get(name)!.push({ score, cx: cx960 });
      }
    }
  }

  // Count copies needed per name
  const nameCounts = new Map<string, number>();
  for (const name of knownNames) {
    nameCounts.set(name, (nameCounts.get(name) ?? 0) + 1);
  }

  // Assign positions with dedup
  const usedCx = new Set<number>();
  for (const [name, needed] of nameCounts) {
    const positions: HandCardPosition[] = [];
    const candidates = (cardMatches.get(name) ?? []).sort((a, b) => b.score - a.score);
    for (const { cx, score } of candidates) {
      if (positions.length >= needed) break;
      if (![...usedCx].some(ux => Math.abs(cx - ux) < 30)) {
        const [adjCx, adjCy] = handAdjust(cx);
        positions.push({ name, cx: adjCx, cy: adjCy, score });
        usedCx.add(cx);
      }
    }
    resultMap.set(name, positions);
  }

  return resultMap;
}

/** Estimate hand card position without OCR. Last resort fallback. */
export function estimateHandPosition(index: number, handSize: number): [number, number] {
  const spacing = Math.min(80, Math.floor(400 / Math.max(handSize, 1)));
  const cx = 480 - Math.floor((handSize - 1) * spacing / 2) + index * spacing;
  return handAdjust(cx);
}
